Add readOnlySegments to store read only segments' indexes.

zhrt123 opened the pull request:
1.length(dtxSegments) < 2: 1PC
2.length(dtxSegments) >= 2: 2PC

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

hyongtao-code commented on src/backend/cdb/cdbtm.c:
```diff
@@ -2370,16 +2372,49 @@ currentGxactWriterGangLost(void)
+	MyTmGxactLocal->readOnlySegmentsMap = bms_add_member(MyTmGxactLocal->readOnlySegmentsMap, segindex);
+
+	MyTmGxactLocal->readOnlySegments = lappend_int(MyTmGxactLocal->readOnlySegments, segindex);
+	
```
Assert(list_length(MyTmGxactLocal->dtxSegments) +
		   list_length(MyTmGxactLocal->readOnlySegments) <= getgpsegmentCount());

zhrt123 replied:
```suggestion
	Assert(list_length(MyTmGxactLocal->dtxSegments) + list_length(MyTmGxactLocal->readOnlySegments) <= getgpsegmentCount());
```

higuoxing commented on src/backend/cdb/cdbtm.c:
```diff
@@ -2370,16 +2372,51 @@ currentGxactWriterGangLost(void)
+	if (segindex == -1)
+		return;
+
+	oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+	/* skip if record already */
+	if (bms_is_member(segindex, MyTmGxactLocal->readOnlySegmentsMap) || 
+		bms_is_member(segindex, MyTmGxactLocal->dtxSegmentsMap))
+	{
+		MemoryContextSwitchTo(oldContext);
+		return;
+	}
```
```suggestion
	/* skip if record already */
	if (bms_is_member(segindex, MyTmGxactLocal->readOnlySegmentsMap) || 
		bms_is_member(segindex, MyTmGxactLocal->dtxSegmentsMap))
	{
		return;
	}
	oldContext = MemoryContextSwitchTo(TopTransactionContext);
```

`bms_is_member()` doesn't call `palloc()`. We don't need to switch the memory context for it, right?

zhrt123 replied:
ok

zhrt123 committed:
Add readOnlySegments to store read only segments' indexes.
1.length(dtxSegments) < 2: 1PC
2.length(dtxSegments) >= 2: 2PC

zhrt123 committed:
transaction commit:
Phase 1: prepare write operations.
Phase 2: commit read-only operations and prepared write operations.

higuoxing commented on src/backend/cdb/cdbtm.c:
```diff
@@ -2388,27 +2461,36 @@ addToGxactDtxSegments(Gang *gang)
-	for (i = 0; i < gang->size; i++)
+
+	/* skip if record already */
+	if (bms_is_member(segindex, MyTmGxactLocal->dtxSegmentsMap))
 	{
-		segdbDesc = gang->db_descriptors[i];
-		Assert(segdbDesc);
-		segindex = segdbDesc->segindex;
+		MemoryContextSwitchTo(oldContext);
```
No need to switch context here.

```
if (bms_is_member())
    return;

oldcontext = MemoryContextSwitchTo();
...
MemoryContextSwitchTo(oldcontext);
```

zhrt123 committed:
performance improve
